### DISCUSSION TITLE
Sparse model generation using lists

### DISCUSSION THREAD

**Discussion by sbmack** [dirkschumacher/ompr#406] [Ideas]
This is related to the recent issue: https://github.com/dirkschumacher/ompr/issues/398#issue-1116308785

A while ago I suggested the possibility using lists to store and process sparse data vectors/matrices.  Here is a simple example of generating a `list` of feasible variable values for a square matrix of "sites":

```
set.seed(7)
# dimension the matrix, specify 0/1 density and generate the dense matrix mat1
sites <- 200
density <- 0.01
vars <- sites * sites
coeff <- sample(0:1, vars, replace = TRUE, prob = c(1 - density, density))
mat1 <- matrix(coeff, nrow = sites)

# convert the matrix mat1 to a list of non-zero value vectors
# list generation is almost instantaneous
intime <- Sys.time()
list1 <- apply(mat1, 1, function(x) which(x > 0))
listtime <- Sys.time() - intime

# rhs is a simple scalar value proportional to density * number of sites
rhs <- round(0.6 * density * sites)

intime <- Sys.time()
model <- MIPModel() %>%
  add_variable(x[i, j], i = 1:sites, j = list1[[i]]) %>%
  set_objective(sum_over(x[i, j], i=1:sites, j = list1[[i]]))  %>% 
  add_constraint(sum_over(x[i, j], j = list1[[i]]) <= rhs, i = 1:sites)

modeltime <- Sys.time() - intime
tottime <- listtime + modeltime

listtime
modeltime
tottime
```
The matrix is randomly filled with 0/1 values with a density of 0.01.  The idea is to create a `list` of sparse vectors maintained as feasible i. j combinations from a dense matrix.  The `list` values are then passed to the MIPModel.  My thinking that under this paradigm, all constraint processing would be done external to MIPModel to minimize the complexity of the formulation inside of MIPModel, since R itself allows for very fast and efficient processing of set operations.  Matrix size for the posted issue was 8,000 by 8,000 so I wanted to test that size as the target.  The model was generated using the ompr developer version 1.0.1.9000.

- Tested a 200 x 200 matrix with density 0.01 yielded approximately 400 feasible pairs.  Generating the model took 8.7 seconds.
- Testing a 1000 x 1000 matrix with density 0.01 yielded approximately 10,000 feasible pairs.  Generating the model took 265 seconds.
- Testing a 4000 x  4000 matrix with density 0.01 yielded approximately 160,000 feasible pairs.  Generation was cancelled without completion after 78 minutes.  So a test with 8,000 sites was moot.

The processing times are not linearly proportional to the number of  feasible variables.  And the approach is obviously not extensible to (very) large matrices.  But it is not clear to me how the processing required by MIPModel could be made any simpler.

It seems to me that the `list` approach is a simple way to deliver sparse data to MIPModel.   However, I can program but I'm not programmer.  So recognize that there may be additional efficiencies available such as parallel processing.  I also realize that there may already be work done related to sparse processing that I am not aware of.

So I am posting this as a discussion item for comments about using lists for data delivery to MIPModel and also appreciate feedback if I am missing something related to maximizing the efficiency of constraint generation inside of MIPModel.

**Comment by dirkschumacher** [comment]
Thanks @sbmack for the experiment. Really appreciated it!

My preferred way to formulate sparse models would be to use filter conditions like the following. Here we check if the coefficient matrix is non-zero on each call.

```R
model <- MIPModel() %>%
    add_variable(x[i, j], i = 1:sites, j = 1:sites, mat1[i, j] > 0) %>%
    set_objective(sum_over(x[i, j], i=1:sites, j = 1:sites, mat1[i, j] > 0))  %>% 
    add_constraint(sum_over(x[i, j], j = 1:sites, mat1[i, j] > 0) <= rhs, i = 1:sites)
```

However upon examining your model, it seems the majority of the time is spent with adding constraints. So there is very likely some potential speed improvements here. To be honest, I am surprised that this takes so long, given the other calls.

``` r
library(ompr)

set.seed(7)
# dimension the matrix, specify 0/1 density and generate the dense matrix mat1
sites <- 300
density <- 0.01
vars <- sites * sites
coeff <- sample(0:1, vars, replace = TRUE, prob = c(1 - density, density))
mat1 <- matrix(coeff, nrow = sites)

# rhs is a simple scalar value proportional to density * number of sites
rhs <- round(0.6 * density * sites)

system.time({
  model <- MIPModel() |>
    add_variable(x[i, j], i = 1:sites, j = 1:sites, mat1[i, j] > 0)
})
#>    user  system elapsed 
#>   0.017   0.001   0.018
system.time({
  model <- set_objective(model, sum_over(x[i, j], i = 1:sites, j = 1:sites, mat1[i, j] > 0))
})
#>    user  system elapsed 
#>   0.052   0.001   0.053
system.time({
  model <- add_constraint(model, sum_over(x[i, j], j = 1:sites, mat1[i, j] > 0) <= rhs, i = 1:sites)
})
#>    user  system elapsed 
#>  12.835   0.503  13.342
```

<sup>Created on 2022-01-31 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by dirkschumacher** [reply]
Ok, so apparently, `add_constraint(model, sum_over(x[i, j], j = 1:sites, mat1[i, j] > 0) <= rhs, i = 1:sites)` spends most of the time waiting for the garbage collector. Let's see if I can improve the memory handling a bit, so that the GC is not triggered.

**Reply by dirkschumacher** [reply]
Ok, I found the problem. {listcomp} was hashing objects as part of generating symbol names. However, I believe that this not actually needed. The hashing allocated memory (plus tons of new symbol names) which then triggered the garbage collector in such a way, that the function spent most of the time garbage collecting memory.

Here is the 4000x4000 model with the new version of [{listcomp}](https://github.com/dirkschumacher/listcomp/pull/11) `remotes::install_github("dirkschumacher/listcomp")`:


``` r
library(ompr)

set.seed(7)
# dimension the matrix, specify 0/1 density and generate the dense matrix mat1
sites <- 4000
density <- 0.01
vars <- sites * sites
coeff <- sample(0:1, vars, replace = TRUE, prob = c(1 - density, density))
mat1 <- matrix(coeff, nrow = sites)

# rhs is a simple scalar value proportional to density * number of sites
rhs <- round(0.6 * density * sites)

system.time({
  model <- MIPModel() |>
    add_variable(x[i, j], i = 1:sites, j = 1:sites, mat1[i, j] > 0) |>
    set_objective(sum_over(x[i, j], i=1:sites, j = 1:sites, mat1[i, j] > 0)) |>
    add_constraint(sum_over(x[i, j], j = 1:sites, mat1[i, j] > 0) <= rhs, i = 1:sites)
})
#>    user  system elapsed 
#>  32.000   0.165  32.220
model
#> Mixed integer linear optimization problem
#> Variables:
#>   Continuous: 160363 
#>   Integer: 0 
#>   Binary: 0 
#> Model sense: maximize 
#> Constraints: 4000
```

<sup>Created on 2022-01-31 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by sbmack** [comment]
Installed ompr version 1.0.2.9000 and ran the scenario with 4000 sites.  Where the test was previously aborted at 78 minutes, the model now is generated in 67 seconds using the list method.  That is great work! 

When processing out the zero values inside of MIPModel as you suggested the execution time was 89 seconds.  So the list method is approximately 25% faster for this test case.

My larger interest in using lists is in generating compound indexes that are often subsetted.  Because in those cases model management can get somewhat messy.  Using the list strategy, the data would be stored in a list of lists, i.e. list hierarchies.  I will follow up with toy model example.

Thanks much for the effort and the reply.

SteveM (sbmack)
